helix-core/src/shellwords.rs

```diff
@@ -1,198 +1,476 @@
+    in_quotes: bool,
+    quote: u8,
+    in_list: bool,
+    is_finished: bool,
+}
+
+impl<'a> Args<'a> {
+    #[inline]
+    fn parse(input: &'a str) -> Self {
+        Self {
+            input,
+            bytes: input.as_bytes(),
+            idx: 0,
+            start: 0,
+            in_quotes: false,
+            quote: b'\0',
+            in_list: false,
+            is_finished: false,
+        }
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.input.is_empty()
+    }
+
+    /// Returns the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::shellwords::Args;
+    /// let args = Args::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!(r#"sed -n "s/test t/not /p""#, args.raw());
+    ///
+    /// let args = Args::from(r#"cat "file name with space.txt""#);
+    /// assert_eq!(r#"cat "file name with space.txt""#, args.raw());
+    /// ```
+    #[inline]
+    pub fn raw(&self) -> &str {
+        self.input
+    }
+
+    #[inline(always)]
+    pub const fn empty() -> Self {
+        Self {
+            input: "",
+            bytes: &[],
+            idx: 0,
+            start: 0,
+            in_quotes: false,
+            quote: b'\0',
+            in_list: false,
+            is_finished: true,
+        }
+    }
+}
+
+#[allow(clippy::copy_iterator)]
+impl<'a> Iterator for Args<'a> {
+    type Item = &'a str;
+
+    #[inline]
+    #[allow(clippy::too_many_lines)]
+    fn next(&mut self) -> Option<Self::Item> {
+        /// Walks backwords checking for `\` and counting them to determine if escaped.
+        #[inline]
+        const fn is_escaped(bytes: &[u8]) -> bool {
+            let mut backslashes = 0;
+            let mut idx = bytes.len() - 1;
+
+            loop {
+                if bytes[idx] != b'\\' {
+                    break;
+                }
+                backslashes += 1;
+
+                idx = match idx.checked_sub(1) {
+                    Some(idx) => idx,
+                    None => break,
+                };
+            }
+
+            (backslashes % 2) != 0
+        }
+
+        if self.is_finished {
+            return None;
+        }
+
+        while self.idx < self.bytes.len() {
+            match self.bytes[self.idx] {
+                b'"' | b'\'' | b'`' if !self.in_list => {
+                    if self.in_quotes {
+                        if self.bytes[self.idx] == self.quote
+                            && !is_escaped(&self.bytes[..self.idx])
+                        {
+                            let arg = Some(&self.input[self.start..self.idx]);
+                            self.in_quotes = false;
+                            self.quote = b'\0';
+                            self.idx += 1;
+                            self.start = self.idx;
+                            return arg;
+                        }
+                        self.idx += 1;
+                    } else if self.idx == self.bytes.len() - 1 {
```

Comment:
Are you referring to the trailing quote example here? I added a comment to a test down below that explained a bit more at the reasoning behind it:
> last (quote) is important, as if the user input an accidental quote at the end, this should be checked in
commands where there should only be one input and return an error rather than silently succeed.

The hope was to try to enforce correct usage of a command, and the quotes here, I believe, are the only thing that is silently "eaten" in such a case.